### ISSUE TITLE
FUTURE: Graph ML layers (Arrow FDD → sklearn → PyG) + ML Lab UI

### ISSUE THREAD

**Issue by bbartling** [bbartling/open-fdd#211]
## Summary

Architect a **three-layer FDD stack** on top of the Arrow-native operator platform. Layer A is production today (deterministic Arrow rules). Layers B and C add graph-aware ML without putting pandas/sklearn/torch inside the Rule Lab sandbox.

Reference concept: [Toy HVAC Graph FDD gist](https://gist.github.com/bbartling/dd7a2a25f1e91b0664dd7fea671c93ce)

**Design principle:** Arrow historian + BRICK model are the **data plane**. Rule engine produces **explainable labels**. ML runs in a **separate training/scoring service** (offline or scheduled batch), not inside `apply_faults_arrow()`.

---

## Layer A — Production (now)

**Status:** Shipped (Arrow-native Rule Lab, feather historian, FDD batch, check-engine).

| Piece | Source | Output |
|-------|--------|--------|
| BACnet poll | `points.csv` / poll driver | `workspace/bacnet/polls/samples.csv` |
| Historian | `feather_store` | PyArrow tables per site/point |
| BRICK model | `model.json` + TTL | equipment tree, `feeds` / `hasPoint` edges |
| FDD rules | `rules_py/*.py` → `apply_faults_arrow` | `fdd_results.json`, `/api/faults/status` |
| Labels (weak) | Rule batch flags per point/window | training labels for Layer B |

**Acceptance (already met):**
- [ ] All enabled rules use `apply_faults_arrow`
- [ ] FDD batch drives traffic light + fault tree
- [ ] No pandas/numpy in Rule Lab sandbox

**Layer A is never replaced by ML** — it remains the explainable engineering layer (flatline, OOB, spread, schedule, etc.).

---

### Goal

Per-equipment **fault likelihood** (0–1) using graph-enriched tabular features and scikit-learn. Compare against Layer A rule outcomes; surface hybrid fault cards in the dashboard.

### Phase B0 — Experiment package (no UI)

**Deliverables**
- `experiments/hvac_graph_fdd/` (port gist `toy_hvac_graph_fdd.py`)
- Read real bench data: feather store + `model.json` for one AHU + N VAVs
- Build adjacency from BRICK `feeds` / `isPartOf` (fallback: manual JSON for bench)
- Feature matrix per node per window (15 / 60 min):
  - rolling mean, min, max, std, slope, last value
  - setpoint error where available
  - rule flag columns from Layer A batch (binary per cookbook pattern)
- Hand-written message passing: `H = concat(X, A@X, A@A@X)` (gist approach)
- Baseline models:
  - `LogisticRegression` (supervised, rule labels)
  - `RandomForestClassifier`
  - `IsolationForest` (unsupervised / normal-only training)

**Acceptance**
- [ ] Script runs on `demo` / `bens-office` bench without Rule Lab changes
- [ ] Documented train/test split + precision/recall on synthetic + rule-labeled windows
- [ ] Export artifact: `workspace/data/ml_models/<site_id>/graph_sklearn_v1.joblib`

### Phase B1 — Feature pipeline (bridge module)

**New modules (proposed)**
- `openfdd_bridge/ml_features.py` — Arrow window rollups → node feature table
- `openfdd_bridge/ml_graph.py` — BRICK → adjacency + node index
- `openfdd_bridge/ml_labels.py` — merge `fdd_results.json` + rule batch into per-node labels

**Data contract**
```python
# NodeFeatureTable (PyArrow)
# columns: window_start, equipment_id, brick_type, feature_*, rule_flag_*, label (optional)

# GraphMeta (JSON)
# nodes: [{equipment_id, point_ids[], brick_class}]
# edges: [{from, to, relation}]  # feeds, serves, hasPoint
```

**Acceptance**
- [ ] `python -m openfdd_bridge.ml_features --site demo --window 60m` writes parquet under `workspace/data/ml_features/`
- [ ] Unit tests: adjacency from bench `model.json`, rolling stats from demo feather

### Phase B2 — Training API + scheduler

**Endpoints (integrator/agent)**
| Method | Path | Purpose |
|--------|------|---------|
| POST | `/api/ml/train` | Train sklearn model for `site_id`, feature set, label source |
| GET | `/api/ml/models` | List artifacts + metadata (version, metrics, trained_at) |
| POST | `/api/ml/score` | Run inference on latest window(s); persist scores |
| GET | `/api/ml/scores/latest` | Per-equipment probabilities for dashboard |
| GET | `/api/ml/compare` | Side-by-side rule vs ML for same window |

**Training job inputs**
```json
{
  "site_id": "demo",
  "model_type": "logistic_regression | random_forest | isolation_forest",
  "label_source": "fdd_batch | manual | synthetic",
  "window_minutes": 60,
  "graph_hops": 2,
  "train_lookback_days": 30
}
```

**Training job outputs** (`workspace/data/ml_models/<site>/manifest.json`)
```json
{
  "version": "graph_sklearn_v3",
  "trained_at": "ISO8601",
  "metrics": {
    "accuracy": 0.92,
    "precision": 0.88,
    "recall": 0.85,
    "f1": 0.86,
    "roc_auc": 0.91,
    "confusion_matrix": [[...],[...]]
  },
  "feature_columns": ["..."],
  "node_count": 12,
  "window_minutes": 60,
  "label_distribution": {"0": 1100, "1": 120}
}
```

**Acceptance**
- [ ] Train + score run in subprocess (like playground worker), not in API worker thread
- [ ] sklearn optional extra: `pip install open-fdd[ml]` or bridge `requirements-ml.txt`
- [ ] Scores written to `workspace/data/ml_scores.json` (or feather) with `source: "ml_graph_sklearn"`

### Phase B3 — Hybrid fault cards (dashboard)

**Fault card merge logic**
```
if rule_fault AND ml_score > threshold → "confirmed" (high confidence)
if rule_fault AND ml_score low      → "rule only" (investigate rule FP)
if no rule AND ml_score > threshold → "ML watch" (anomaly, needs review)
if both agree normal                → no card
```

**UI touchpoints (existing pages)**
- **Faults / check-engine:** badge `ML 0.91` next to rule code; expand evidence panel
- **Equipment graphic:** color ring by max(rule severity, ml tier)
- **Rule Lab:** read-only panel "ML agreement rate for this rule (7d)" — no training here

**Acceptance**
- [ ] `/api/faults/status` includes optional `ml_scores[]` when model deployed
- [ ] Operator sees evidence bullets (gist-style), not raw probabilities alone

### Phase B4 — Operational cadence (Layer B)

| Job | Default cadence | Notes |
|-----|-----------------|-------|
| **Feature materialization** | Every poll ingest (or hourly) | Append windows to `ml_features/` |
| **Inference (score)** | Every **15 min** (align with shortest poll) or hourly on edge | Lightweight: load `.joblib`, score latest window only |
| **Retrain** | **Weekly** (Sunday 02:00 site TZ) or on-demand | Full lookback; compare metrics to prior version |
| **Model promotion** | Manual integrator approve OR auto if `f1 >= prior - 0.02` | Keep previous artifact for rollback |
| **Rule vs ML compare report** | Daily | JSON summary: agreement %, rule-only, ML-only counts |

**Persistence rules**
- Require **3 consecutive windows** above threshold before ML-only fault (mirror FDD persistence)
- ML never auto-writes BACnet; read-only scoring

**Acceptance**
- [ ] `systemd` unit or `workspace-cron` job: `openfdd-ml-score`, `openfdd-ml-train-weekly`
- [ ] Metrics retained 90 days for UI charts

---

### Goal

Graph neural network over **time-encoded node features** (not single snapshot). Target: better upstream/downstream fault separation (AHU static vs single VAV).

### Phase C0 — Temporal feature encoder

- Replace hand message-passing with rolling **sequence** per node: last K windows (e.g. K=12 @ 5 min = 1 h)
- Encoder options (incremental):
  1. Flatten K×F → sklearn (baseline)
  2. 1D CNN / GRU on sequence → embedding
  3. **PyTorch Geometric** `GraphSAGE` or `GAT` on node embeddings

**Dependencies:** `torch`, `torch-geometric` — **never** in bridge default image; separate `openfdd-ml-gnn` container or host venv.

### Phase C1 — GNN training pipeline

- Export training tensors from Arrow feature store
- Train/val split **by time** (no random row shuffle)
- Metrics: per-node AUROC, per-fault-type confusion, calibration plot
- Artifact: `graph_gnn_v1.pt` + ONNX export optional for edge inference

### Phase C2 — GNN inference + hybrid stack

- Score cadence same as Layer B (15 min inference)
- Retrain **monthly** (GNN needs more data; weekly only if GPU host + label volume sufficient)
- Three-way compare: **Rule vs sklearn vs GNN** on `/api/ml/compare`

**Acceptance**
- [ ] Bench demo: AHU-static scenario ranks AHU high, VAV-local scenario ranks single VAV high
- [ ] Inference CPU budget < 30 s per site on Pi 5 / bensserver class hardware

---

## New UI tab: **ML Lab** (`/ml-lab`)

Revive/plan the legacy `MlLabPage` concept as an **integrator-only** tab (not operator). Purpose: training visibility, not live OT control.

### Proposed sections

| Section | Content |
|---------|---------|
| **Overview** | Active model version, last train/score times, site selector, promote/rollback |
| **Training run** | Trigger train, pick model type, label source, lookback days |
| **Metrics dashboard** | Loss curve (GNN), accuracy/F1/ROC-AUC (sklearn), **RMSE/MAE** if regression head added later |
| **Confusion matrix** | Heatmap per fault type / per equipment class |
| **Calibration** | Predicted prob vs observed rate (reliability diagram) |
| **Feature importance** | RF/SKLearn `feature_importances_` or SHAP (optional) |
| **Compare** | Time series: rule faults vs ML scores vs GNN scores (stacked chart) |
| **Data health** | Window count, label balance, missing features %, graph node coverage |
| **Artifacts** | List `.joblib` / `.pt` files, download, delete (integrator) |

### Metrics by layer

| Layer | Primary metrics in UI |
|-------|----------------------|
| B sklearn classifier | accuracy, precision, recall, F1, ROC-AUC, confusion matrix |
| B isolation forest | anomaly rate %, score distribution, contamination param |
| C GNN classifier | train/val **loss**, accuracy, per-class F1, AUROC |
| C GNN regressor (future) | **RMSE**, **MAE**, R² on setpoint error / residual prediction |

### Role gating

- **Operator:** see ML badges on Faults page only (no ML Lab tab)
- **Integrator / agent:** full ML Lab + train/score triggers

---

## End-to-end architecture (target)

```
BACnet poll → feather (Arrow)
                 │
                 ├─► FDD batch (Layer A) ──► rules faults / labels
                 │
                 └─► ml_features (windows)
                          │
                     BRICK graph
                          │
              ┌───────────┴───────────┐
              ▼                       ▼
        sklearn train/score     GNN train/score (Layer C)
              │                       │
              └───────────┬───────────┘
                          ▼
              ml_scores + compare API
                          ▼
              Faults UI + ML Lab tab
```

---

## Suggested implementation order (turnkey checklist)

- [ ] **B0** — `experiments/hvac_graph_fdd` on bench + gist parity tests
- [ ] **B1** — Arrow feature + graph builders in bridge
- [ ] **B2** — `/api/ml/*` + artifacts + cron score/train
- [ ] **B3** — Hybrid fault cards on Faults page
- [ ] **B4** — ML Lab tab (metrics, compare, train trigger)
- [ ] **C0–C2** — Temporal encoder + PyG (separate optional image)
- [ ] **Docs** — `docs/developer/ml-graph-fdd.md` + operator glossary entry

---

## Open questions

1. **Label source of truth:** Rule batch only, or allow operator ack/shelve as negative labels?
2. **Auto-promote models** on edge without integrator review — default off?
3. **GNN inference** on Pi edge vs train-on-server / score-on-server only?
4. **Regression vs classification** first for sklearn (classification matches gist; RMSE applies if predicting airflow residual)?

---

## Related

- Gist: https://gist.github.com/bbartling/dd7a2a25f1e91b0664dd7fea671c93ce
- Skill stub: `skills/ml-lab-sklearn/SKILL.md`
- Example notebook: `examples/AHU/RTU7_machine_learning.ipynb`
- Arrow runtime: `docs/developer/arrow-native-runtime.md`
